plugin/azure/azure.go

```diff
@@ -0,0 +1,227 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/file"
+	"github.com/coredns/coredns/plugin/pkg/fall"
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+	"github.com/coredns/coredns/request"
+
+	azure "github.com/Azure/azure-sdk-for-go/profiles/latest/dns/mgmt/dns"
+	"github.com/miekg/dns"
+)
+
+type zone struct {
+	id  string
+	z   *file.Zone
+	dns string
+}
+
+type zones map[string][]*zone
+
+// Azure is the core struct of the azure plugin
+type Azure struct {
+	Next      plugin.Handler
+	Fall      fall.F
+	zoneNames []string
+	client    azure.RecordSetsClient
+	upstream  *upstream.Upstream
+	zMu       sync.RWMutex
+	zones     zones
+}
+
+// New validates the input dns zones and initializes the Azure struct
+func New(ctx context.Context, dnsClient azure.RecordSetsClient, keys map[string][]string, up *upstream.Upstream) (*Azure, error) {
+	zones := make(map[string][]*zone, len(keys))
+	zoneNames := make([]string, 0, len(keys))
+	for resourceGroup, inputZoneNames := range keys {
+		for _, zoneName := range inputZoneNames {
+			_, err := dnsClient.ListAllByDNSZone(context.Background(), resourceGroup, zoneName, nil, "")
+			if err != nil {
+				return nil, err
+			}
+			zoneNameFQDN := zoneName
+			// Normalizing zoneName to make it fqdn if required
+			if !strings.HasSuffix(zoneNameFQDN, ".") {
+				zoneNameFQDN = fmt.Sprintf("%s.", zoneNameFQDN)
+			}
+			if _, ok := zones[zoneNameFQDN]; !ok {
+				zoneNames = append(zoneNames, zoneNameFQDN)
+			}
+			zones[zoneNameFQDN] = append(zones[zoneNameFQDN], &zone{id: resourceGroup, dns: zoneName, z: file.NewZone(zoneName, "")})
+		}
+	}
+	return &Azure{
+		client:    dnsClient,
+		zones:     zones,
+		zoneNames: zoneNames,
+		upstream:  up,
+	}, nil
+}
+
+// Run updates the zones once and starts a goroutine to do it every minute
+func (h *Azure) Run(ctx context.Context) error {
+	if err := h.updateZones(ctx); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Infof("Breaking out of Azure update loop: %v", ctx.Err())
+				return
+			case <-time.After(1 * time.Minute):
+				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil {
+					log.Errorf("Failed to update zones: %v", err)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (h *Azure) updateZones(ctx context.Context) error {
+	errc := make(chan error)
+	defer close(errc)
+	for zName, z := range h.zones {
+		go func(zName string, z []*zone) {
```

Comment:
Is it important that these zones are updated in a go-routine each, or can update them in sequence (and so away with all the channel business here)

Reply:
this is inspired by `route53` which uses multiple goroutines to update the zones.

Reply:
I think this can be just sequential

Reply:
sure will do, should I make it sequential for route53 too?